I am aware that gcc and clang are not linkers (they are wrappers around preprocessing, compiling, and linking). From what I know they can be used as linkers and call `ld` with appropriate arguments. Here is the error if I do not manually set `cc` as the linker. 

```
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c seq.c -o .objs/seq.o
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c symtable.c -o .objs/symtable.o
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c util.c -o .objs/util.o
gcc -I.. -Wall -Wextra -std=c11 -pedantic -O2 -g -fPIC -c values.c -o .objs/values.o
ld -shared -Wl,-soname,libndtypes.so.0 -o libndtypes.so.0.2.0dev3 .objs/alloc.o .objs/attr.o .objs/context.o .objs/copy.o .objs/encodings.o .objs/equal.o .objs/grammar.o .objs/io.o .objs/lexer.o .objs/match.o .objs/ndtypes.o .objs/parsefuncs.o .objs/parser.o .objs/seq.o .objs/symtable.o .objs/util.o .objs/values.o compat/.objs/bpgrammar.o compat/.objs/bplexer.o compat/.objs/import.o
/nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/ld: unrecognized option '-Wl,-soname,libndtypes.so.0'
/nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/ld: use the --help option for usage information
make[1]: *** [Makefile:57: libndtypes.so.0.2.0dev3] Error 1
make[1]: Leaving directory '/build/source/libndtypes'
make: *** [Makefile:23: default] Error 2
builder for '/nix/store/5gplhcy8rxmiabvw43agnwpkflrvz29l-libndtypes-0.2.0dev3.drv' failed with exit code 2
error: build of '/nix/store/5gplhcy8rxmiabvw43agnwpkflrvz29l-libndtypes-0.2.0dev3.drv' failed
```

`'-Wl,-soname,libndtypes.so.0'` is a argument that is passed to `gcc` or `clang` that is appropriately passed to `ld`.

--- a/pkgs/development/libraries/libndtypes/default.nix
+++ b/pkgs/development/libraries/libndtypes/default.nix
@@ -14,7 +14,9 @@ stdenv.mkDerivation rec {
     sha256 = "0dpvv13mrid8l5zkjlz18qvirz3nr0v98agx9bcvkqbiahlfgjli";
   };
 
-  makeFlags = [ "CONFIGURE_LDFLAGS='-shared'" ];
+  # Override linker with cc (symlink to either gcc or clang)
+  # Library expects to use cc for linking
+  configureFlags = [ "LD=${stdenv.cc.targetPrefix}cc" ];
 
   meta = {
     description = "Dynamic types for data description and in-memory computations";

--- a/pkgs/development/libraries/libxnd/default.nix
+++ b/pkgs/development/libraries/libxnd/default.nix
@@ -17,10 +17,9 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ libndtypes ];
 
-  configureFlags = [ "XND_INCLUDE='-I${libndtypes}/include'"
-                     "XND_LINK='-L${libndtypes}/lib'" ];
-
-  makeFlags = [ "CONFIGURE_LDFLAGS='-shared'" ];
+  # Override linker with cc (symlink to either gcc or clang)
+  # Library expects to use cc for linking
+  configureFlags = [ "LD=${stdenv.cc.targetPrefix}cc" ];
 
   meta = {
     description = "General container that maps a wide range of Python values directly to memory";

